Asynchronous process reaping

takase1121 opened the pull request:
This PR allows process handles to be reaped asynchronously.
Previously, process reaping on `__gc` will stall the GC due to waiting for process to actually exit.
Now, this PR will allow the GC to run continuously while the process is sent to a background thread where it is terminated.

This also fixes an issue on Windows where console applications are improperly kept open, holding lite-xl in the background.

Guldoman commented on src/api/process.c:
```diff
@@ -21,11 +22,16 @@
 #endif
 
 #define READ_BUF_SIZE 2048
+#define KILL_LIST_MIN_SIZE 16
```
Is this used somewhere?

takase1121 replied:
Nope, old code that I'll remove once I got time.

Guldoman commented on src/api/process.c:
```diff
@@ -542,11 +731,16 @@ static int f_running(lua_State* L) {
 }
 
-static const struct luaL_Reg lib[] = {
+static int process_gc(lua_State *L) {
+  kill_list_wait_all(&kill_list);
+  SDL_WaitThread(kill_list_thread, NULL);
+  kill_list_free(&kill_list);
+  return 0;
+}
```
Instead of doing this in Lua gc, we could do this in `main` and close the window in the meantime, right?

takase1121 replied:
I want to keep process stuff in Process. In my testing, it doesn't halt the closure of the window for too long.

Of course, a better fix is to decouple the window.

Guldoman commented on src/api/process.c:
```diff
@@ -38,9 +44,24 @@ typedef struct {
+typedef struct {
+  bool stop;
+  SDL_mutex *mutex;
+  SDL_cond *has_work, *work_done;
```
What is `work_done` used for?

takase1121 replied:
It is used to signal work done.
If a process is abandoned or killed, it will be signaled. The waiter waits until the list is empty.

Guldoman replied:
I don't think it's ever signaled tho.

Guldoman commented on src/api/process.c:
```diff
@@ -63,60 +84,205 @@ typedef enum {
+      if ((SDL_GetTicks() - current_task->start_time) < PROCESS_TERM_DELAY) {
+        last_task = current_task;
+        current_task = current_task->next;
+        continue;
```
If the head task hasn't expired, aren't all other tasks the same?
What I mean is that tasks are added by a single thread all with the same timeout, and are always added to the tail. So the first one that could possibly expire is the first one, right?

Also what do you think about using `SDL_Delay` with the time until the first task?
Basically:
1. Remove task from list
2. Check timeout
3. If timed out, goto 7
4. Unlock mutex
5. SDL_Delay for needed time
6. Lock mutex
7. Sigterm it, then add it back to the queue with new timeout, or sigkill it if out of retries
8. Goto 1

takase1121 replied:
I would assume that will wait for let's say 50ms per task.
In the meantime, you can have 10 tasks all completed in the 50ms wait period.

Instead of waiting once per task, it would be more efficient to wait once and sweep.
I'll agree with the first part, which I'll be fixing by breaking out of the loop.

Another method is process the tasks then enter an alertable sleep.
If this sleep completed, it will be stopped by the CondWait at the top.
Otherwise, it'll bypass the CondWait on the top and go straight to processing stuff.
This should make the thread "less spinny" when there's nothing.

Guldoman commented on src/api/process.c:
```diff
@@ -525,14 +691,37 @@ static int self_signal(lua_State* L, signal_e sig) {
 static int f_kill(lua_State* L) { return self_signal(L, SIGNAL_KILL); }
 static int f_interrupt(lua_State* L) { return self_signal(L, SIGNAL_INTERRUPT); }
+
 static int f_gc(lua_State* L) { 
```
Little bit of trailing whitespace from some older commit.

Guldoman commented on src/api/process.c:
```diff
@@ -562,12 +756,30 @@ static const struct luaL_Reg lib[] = {
+};
+
 int luaopen_process(lua_State *L) {
+  setvbuf(stdout, NULL, _IONBF, 0);
```
What's this for?

takase1121 replied:
Debugging. Will remove.

Guldoman commented on src/api/process.c:
```diff
@@ -63,60 +84,210 @@ typedef enum {
+
+    if (list->stop) break;
+
+    for (i = 0, size = list->size; i < size; i++) {
```
Instead of this `for`, could we use something like `while (current_task)`?
Just a nit, but I feel like it would be a bit cleaner this way.

takase1121 replied:
The whole reason size was added into the struct was because you can't simply use `while (current_task)` if you want to push to the same list. It has potential to deadlock (or at least spinlock until all tasks completed).

I've considered tracking list->tail and checking that, but in the end settled for this solution because it is the simplest to understand and implement.

Guldoman commented on src/api/process.c:
```diff
@@ -63,60 +84,210 @@ typedef enum {
+    // if the queue is empty, we can immediately wait for another task or to quit.
+    if (list->size) {
+      SDL_UnlockMutex(list->mutex);
+      SDL_Delay(PROCESS_TERM_DELAY);
```
Could we delay the amount of time needed for the task to timeout?
This way if a task only needed 10ms to timeout, we wouldn't waste 40ms waiting.

Guldoman commented on src/api/process.c:
```diff
@@ -63,60 +83,201 @@ typedef enum {
+
+    if (list->stop) break;
+
+    while ((current_task = list->head)) {
```
Last nit, is the double parenthesis intended?

takase1121 replied:
Yes, `current_task = list->head` is a statement and `(current_task = list->head)` is an expression.

Guldoman reviewed (commented):
I still need to test this, but code-wise lgtm.

Guldoman commented on src/api/process.c:
```diff
@@ -63,60 +83,201 @@ typedef enum {
+        free(current_task);
+      }
+    }
+    delay = list->head ? SDL_GetTicks() - list->head->start_time : 0;
```
The delay might be a bit wrong here.
Does
```suggestion
    delay = list->head ? (list->head->start_time + PROCESS_TERM_DELAY) - SDL_GetTicks() : 0;
```
look more correct?

takase1121 commented:
Tested on macOS Monterey:

<details>
<summary>Skipping SIGTERM before adding to the kill list</summary>

```
Process 993 launched: '/Users/takase/gh/lite-xl/lite-xl' (x86_64)
(lldb)  jump 698
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $132 = 13
Added (1005)
(lldb)  jump 698
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $133 = 17
SIGTERM 1 (1005)
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $134 = 13
Added (1004)
(lldb)  jump 698
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $135 = 13
Added (1003)
(lldb)  jump 698
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $136 = 14
Free 2 (1005)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $137 = 17
SIGTERM 1 (1004)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $138 = 17
SIGTERM 1 (1003)
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $139 = 13
Added (1002)
(lldb)  jump 698
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $140 = 13
Added (1001)
(lldb)  jump 698
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $141 = 14
Free 2 (1004)
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $142 = 14
Free 2 (1003)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $143 = 17
SIGTERM 1 (1002)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $144 = 17
SIGTERM 1 (1001)
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $145 = 13
Added (1000)
(lldb)  jump 698
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $146 = 14
Free 2 (1002)
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $147 = 14
Free 2 (1001)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $148 = 17
SIGTERM 1 (1000)
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $149 = 12
Added (999)
(lldb)  jump 698
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $150 = 12
Added (998)
(lldb)  jump 698
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $151 = 14
Free 2 (1000)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $152 = 16
SIGTERM 1 (999)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $153 = 16
SIGTERM 1 (998)
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $154 = 12
Added (997)
(lldb)  jump 698
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $155 = 13
Free 2 (999)
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $156 = 13
Free 2 (998)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $157 = 16
SIGTERM 1 (997)
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $158 = 12
Added (996)
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $159 = 13
Free 2 (997)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $160 = 16
SIGTERM 1 (996)
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $161 = 13
Free 2 (996)
(lldb)  expr printf("Kill list freed\n")
(int) $162 = 16
Kill list freed
```
</details>

<details>
<summary>Running a script that traps SIGTERM</summary>

```
Process 1219 launched: '/Users/takase/gh/lite-xl/build/src/lite-xl' (x86_64)
2023-03-22 16:40:15.400521+0800 lite-xl[1219:24678] SecTaskLoadEntitlements failed error=22 cs_flags=20, pid=1219
2023-03-22 16:40:15.400661+0800 lite-xl[1219:24678] SecTaskCopyDebugDescription: lite-xl[1219]/0#-1 LF=0
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $346 = 13
Added (1239)
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $347 = 13
Added (1236)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $348 = 17
SIGTERM 1 (1239)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $349 = 17
SIGTERM 1 (1236)
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $350 = 13
Added (1235)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $351 = 17
SIGTERM 2 (1239)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $352 = 17
SIGTERM 2 (1236)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $353 = 17
SIGTERM 1 (1235)
(lldb)  expr printf("SIGKILL %d (%ld)\n", current_task->tries, current_task->handle)
(int) $354 = 17
SIGKILL 3 (1239)
(lldb)  expr printf("SIGKILL %d (%ld)\n", current_task->tries, current_task->handle)
(int) $355 = 17
SIGKILL 3 (1236)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $356 = 17
SIGTERM 2 (1235)
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $357 = 14
Free 4 (1239)
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $358 = 14
Free 4 (1236)
(lldb)  expr printf("SIGKILL %d (%ld)\n", current_task->tries, current_task->handle)
(int) $359 = 17
SIGKILL 3 (1235)
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $360 = 13
Added (1231)
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $361 = 13
Added (1230)
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $362 = 13
Added (1228)
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $363 = 14
Free 4 (1235)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $364 = 17
SIGTERM 1 (1231)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $365 = 17
SIGTERM 1 (1230)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $366 = 17
SIGTERM 1 (1228)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $367 = 17
SIGTERM 2 (1231)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $368 = 17
SIGTERM 2 (1230)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $369 = 17
SIGTERM 2 (1228)
(lldb)  expr printf("SIGKILL %d (%ld)\n", current_task->tries, current_task->handle)
(int) $370 = 17
SIGKILL 3 (1231)
(lldb)  expr printf("SIGKILL %d (%ld)\n", current_task->tries, current_task->handle)
(int) $371 = 17
SIGKILL 3 (1230)
(lldb)  expr printf("SIGKILL %d (%ld)\n", current_task->tries, current_task->handle)
(int) $372 = 17
SIGKILL 3 (1228)
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $373 = 14
Free 4 (1231)
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $374 = 14
Free 4 (1230)
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $375 = 14
Free 4 (1228)
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $376 = 13
Added (1226)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $377 = 17
SIGTERM 1 (1226)
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $378 = 13
Added (1224)
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $379 = 13
Added (1223)
(lldb)  expr printf("Added (%ld)\n", self->pid)
(int) $380 = 13
Added (1222)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $381 = 17
SIGTERM 2 (1226)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $382 = 17
SIGTERM 1 (1224)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $383 = 17
SIGTERM 1 (1223)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $384 = 17
SIGTERM 1 (1222)
(lldb)  expr printf("SIGKILL %d (%ld)\n", current_task->tries, current_task->handle)
(int) $385 = 17
SIGKILL 3 (1226)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $386 = 17
SIGTERM 2 (1224)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $387 = 17
SIGTERM 2 (1223)
(lldb)  expr printf("SIGTERM %d (%ld)\n", current_task->tries, current_task->handle)
(int) $388 = 17
SIGTERM 2 (1222)
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $389 = 14
Free 4 (1226)
(lldb)  expr printf("SIGKILL %d (%ld)\n", current_task->tries, current_task->handle)
(int) $390 = 17
SIGKILL 3 (1224)
(lldb)  expr printf("SIGKILL %d (%ld)\n", current_task->tries, current_task->handle)
(int) $391 = 17
SIGKILL 3 (1223)
(lldb)  expr printf("SIGKILL %d (%ld)\n", current_task->tries, current_task->handle)
(int) $392 = 17
SIGKILL 3 (1222)
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $393 = 14
Free 4 (1224)
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $394 = 14
Free 4 (1223)
(lldb)  expr printf("Free %d (%ld)\n", current_task->tries, current_task->handle)
(int) $395 = 14
Free 4 (1222)
(lldb)  expr printf("Kill list freed\n")
(int) $596 = 16
Kill list freed
```
</details>

<details>
<summary>Disabling SIGKILL</summary>

```
(int) $10 = 20
Abandoning 4 (1720)
(lldb)  expr printf("Abandoning %d (%ld)\n", current_task->tries, current_task->handle)
(int) $11 = 20
Abandoning 4 (1718)
(lldb)  expr printf("Abandoning %d (%ld)\n", current_task->tries, current_task->handle)
(int) $12 = 20
Abandoning 4 (1716)
(lldb)  expr printf("Abandoning %d (%ld)\n", current_task->tries, current_task->handle)
(int) $13 = 20
Abandoning 4 (1715)
(lldb)  expr printf("Abandoning %d (%ld)\n", current_task->tries, current_task->handle)
(int) $14 = 20
Abandoning 4 (1713)
(lldb)  expr printf("Abandoning %d (%ld)\n", current_task->tries, current_task->handle)
(int) $15 = 20
Abandoning 4 (1710)
(lldb)  expr printf("Abandoning %d (%ld)\n", current_task->tries, current_task->handle)
(int) $16 = 20
Abandoning 4 (1708)
(lldb)  expr printf("Abandoning %d (%ld)\n", current_task->tries, current_task->handle)
(int) $17 = 20
Abandoning 4 (1706)
(lldb)  expr printf("Abandoning %d (%ld)\n", current_task->tries, current_task->handle)
(int) $18 = 20
Abandoning 4 (1705)
(lldb)  expr printf("Abandoning %d (%ld)\n", current_task->tries, current_task->handle)
(int) $19 = 20
Abandoning 4 (1704)
(lldb)  expr printf("Kill list freed\n")
(int) $20 = 16
Kill list freed
```

`ps -ax | grep sleep`

```
takase@Takases-MacBook-Pro lite-xl % ps -ax | grep sleep
 1652 ttys000    0:00.00 sleep 1000
 1654 ttys000    0:00.00 sleep 1000
 1655 ttys000    0:00.00 sleep 1000
 1657 ttys000    0:00.00 sleep 1000
 1660 ttys000    0:00.00 sleep 1000
 1662 ttys000    0:00.00 sleep 1000
 1664 ttys000    0:00.00 sleep 1000
 1666 ttys000    0:00.00 sleep 1000
 1667 ttys000    0:00.00 sleep 1000
 1680 ttys000    0:00.00 sleep 1000
 1683 ttys000    0:00.00 sleep 1000
 1684 ttys000    0:00.00 sleep 1000
 1686 ttys000    0:00.00 sleep 1000
 1688 ttys000    0:00.00 sleep 1000
 1690 ttys000    0:00.00 sleep 1000
 1692 ttys000    0:00.00 sleep 1000
 1694 ttys000    0:00.00 sleep 1000
 1695 ttys000    0:00.00 sleep 1000
 1696 ttys000    0:00.00 sleep 1000
 1707 ttys000    0:00.00 sleep 1000
 1709 ttys000    0:00.00 sleep 1000
 1711 ttys000    0:00.00 sleep 1000
 1712 ttys000    0:00.00 sleep 1000
 1714 ttys000    0:00.00 sleep 1000
 1717 ttys000    0:00.00 sleep 1000
 1719 ttys000    0:00.00 sleep 1000
 1721 ttys000    0:00.00 sleep 1000
 1722 ttys000    0:00.00 sleep 1000
 1723 ttys000    0:00.00 sleep 1000
 1729 ttys002    0:00.00 grep sleep
```
</details>
